src/main/java/org/zalando/nakadi/config/AuthenticationConfig.java

```diff
@@ -2,43 +2,127 @@
+        final PoolingHttpClientConnectionManager result = new PoolingHttpClientConnectionManager();
+        result.setMaxTotal(20);
+        result.setDefaultMaxPerRoute(10);
+        return result;
+    }
+
+    @Bean
+    public RequestConfig requestConfig() {
+        final RequestConfig result = RequestConfig.custom()
+                .setConnectionRequestTimeout(2000)
+                .setConnectTimeout(1000)
+                .setSocketTimeout(2000)
+                .build();
+        return result;
+    }
+
+    @Bean
+    public CloseableHttpClient httpClient(final PoolingHttpClientConnectionManager poolingHttpClientConnectionManager,
+                                          final RequestConfig requestConfig) {
+        final CloseableHttpClient result = HttpClientBuilder
+                .create()
+                .setConnectionManager(poolingHttpClientConnectionManager)
+                .setDefaultRequestConfig(requestConfig)
+                .build();
+        return result;
+    }
+
+    @Bean
+    public RestTemplate restTemplate(final HttpClient httpClient) {
+        final HttpComponentsClientHttpRequestFactory requestFactory = new HttpComponentsClientHttpRequestFactory();
+        requestFactory.setHttpClient(httpClient);
+        final RestTemplate restTemplate = new RestTemplate(requestFactory);
+        restTemplate.setErrorHandler(TokenResponseErrorHandler.getDefault());
+        return restTemplate;
+    }
+
+    public static class MeasureAndDispatchResourceServerTokenServices implements ResourceServerTokenServices {
+
         private final Timer timer;
+        private final TokenInfoResourceServerTokenServices remoteService;
+        private final TokenInfoResourceServerTokenServices localService;
+        private final FeatureToggleService featureToggleService;
 
-        public MeasuringTokenInfoResourceServerTokenServices(final String tokenInfoEndpointUrl, final String clientId,
-                                                             final MetricRegistry metricRegistry) {
-            super(tokenInfoEndpointUrl, clientId);
+        public MeasureAndDispatchResourceServerTokenServices(final MetricRegistry metricRegistry,
+                                                             final SecuritySettings securitySettings,
+                                                             final RestTemplate restTemplate,
+                                                             final FeatureToggleService featureToggleService) {
+            remoteService = new TokenInfoResourceServerTokenServices(
+                    securitySettings.getTokenInfoUrl(),
+                    securitySettings.getClientId(),
+                    new DefaultAuthenticationExtractor(),
+                    new DefaultUserRolesProvider(),
+                    restTemplate);
+            localService = new TokenInfoResourceServerTokenServices(
+                    securitySettings.getTokenInfoUrl(),
```

Comment:
@adyach shouldn't this be different from the remote url?

Reply:
fixed